Framework/Muon/src/MuonGroupingAsymmetry.cpp

```diff
@@ -0,0 +1,321 @@
+                  "asymmetry calculation.");
+
+  declareProperty("GroupName", emptyString,
+                  "The name of the group. Must "
+                  "contain at least one alphanumeric "
+                  "character.",
+                  Direction::Input);
+
+  declareProperty(make_unique<ArrayProperty<int>>(
+                      "Grouping", defaultGrouping,
+                      IValidator_sptr(new NullValidator), Direction::Input),
+                  "The grouping of detectors, comma separated list of detector "
+                  "IDs or hyphenated ranges of IDs.");
+
+  declareProperty("AsymmetryTimeMin", 0.0,
+                  "Start time for the asymmetry estimation (in micro "
+                  "seconds). Defaults to the start time of the InputWorkspace.",
+                  Direction::Input);
+
+  declareProperty("AsymmetryTimeMax", 32.0,
+                  "End time for the asymmetry estimation (in micro seconds). "
+                  "Defaults to the end time of the InputWorkspace.",
+                  Direction::Input);
+
+  declareProperty("NormalizationIn", 0.0,
+                  "If this value is non-zero then this is used for the "
+                  "normalization, instead of being estimated.",
+                  Direction::Input);
+
+  declareProperty(make_unique<ArrayProperty<int>>(
+                      "SummedPeriods", defaultPeriods,
+                      IValidator_sptr(new NullValidator), Direction::Input),
+                  "A list of periods to sum in multiperiod data.");
+  declareProperty(
+      make_unique<ArrayProperty<int>>("SubtractedPeriods", Direction::Input),
+      "A list of periods to subtract in multiperiod data.");
+
+  // Perform Group Associations.
+
+  std::string groupingGrp("Grouping Information");
+  setPropertyGroup("GroupName", groupingGrp);
+  setPropertyGroup("Grouping", groupingGrp);
+  setPropertyGroup("AsymmetryTimeMin", groupingGrp);
+  setPropertyGroup("AsymmetryTimeMax", groupingGrp);
+
+  std::string periodGrp("Multi-period Data");
+  setPropertyGroup("SummedPeriods", periodGrp);
+  setPropertyGroup("SubtractedPeriods", periodGrp);
+}
+
+std::map<std::string, std::string> MuonGroupingAsymmetry::validateInputs() {
+  std::map<std::string, std::string> errors;
+
+  std::string groupName = this->getProperty("GroupName");
+  if (groupName.empty()) {
+    errors["GroupName"] = "Group name must be specified.";
+  }
+
+  if (!std::all_of(std::begin(groupName), std::end(groupName),
+                   Mantid::MuonAlgorithmHelper::isAlphanumericOrUnderscore)) {
+    errors["GroupName"] =
+        "The group name must contain alphnumeric characters and _ only.";
+  }
+
+  WorkspaceGroup_sptr inputWS = getProperty("InputWorkspace");
+  std::vector<int> summedPeriods = getProperty("SummedPeriods");
+  std::vector<int> subtractedPeriods = getProperty("SubtractedPeriods");
+
+  WorkspaceGroup_sptr inputGroup =
+      boost::dynamic_pointer_cast<WorkspaceGroup>(inputWS);
```

Comment:
`inputWS` is already declared as `WorkspaceGroup_sptr` above, so why this cast? Am I missing something here?